atoma-state/src/handlers.rs

```diff
@@ -766,6 +767,77 @@ pub(crate) async fn handle_state_manager_event(
     Ok(())
 }
 
+/// Handles an update to the number of compute units in a stack.
+///
+/// This function processes an update to the number of compute units in a stack by parsing the event data
+/// and updating the corresponding stack in the database. If the ratio of compute units is greater than or equal to 95%,
+/// it will submit a claim funds for stacks transaction.
+///
+/// # Arguments
+///
+/// * `state_manager` - A reference to the `AtomaStateManager` for database operations.
+/// * `stack_small_id` - The unique identifier of the stack to be updated.
+/// * `estimated_total_compute_units` - The estimated total number of compute units in the stack.
+/// * `total_compute_units` - The total number of compute units in the stack.
+///
+/// # Returns
+///
+/// * `Result<()>` - Ok(()) if the update was successful, or an error if something went wrong.
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// * The database operation to update the stack number of compute units fails.
+/// * The Sui client operation to submit the claim funds for stacks transaction fails.
+#[instrument(
+    level = "info",
+    skip_all,
+    fields(
+        stack_small_id,
+        estimated_total_compute_units,
+        total_compute_units,
+        ratio
+    )
+)]
+#[allow(clippy::cast_sign_loss)]
+pub(crate) async fn handle_update_stack_num_compute_units_and_claim_funds(
+    state_manager: &AtomaStateManager,
+    stack_small_id: i64,
+    estimated_total_compute_units: i64,
+    total_compute_units: i64,
+) -> Result<()> {
+    info!(
+        target = "atoma-state-handlers",
+        event = "handle-update-stack-num-compute-units-and-claim-funds",
+        "Processing update stack num compute units and claim funds"
+    );
+    let (ratio, stack_computed_units) = state_manager
+        .state
+        .update_stack_num_compute_units(
+            stack_small_id,
+            estimated_total_compute_units,
+            total_compute_units,
+        )
+        .await?;
+    if ratio >= RATIO_FOR_CLAIM_STACK_THRESHOLD {
+        state_manager
```

Comment:
IMHO This is not as easy as doing this.
You can have more inferences for the same stack. You have to put that stack into a lock mode so you don't start any other inference. And after the last is done, then you can claim the stack.